src/lib.rs

```diff
@@ -3789,6 +3801,448 @@ pub struct SemanticHighlightingParams {
+ */
+#[derive(Debug, Eq, PartialEq, Hash, PartialOrd, Clone, Deserialize, Serialize)]
+#[cfg(feature = "proposed")]
+pub struct SemanticTokenModifier(Cow<'static, str>);
+
+#[cfg(feature = "proposed")]
+impl SemanticTokenModifier {
+    pub const DOCUMENTATION: SemanticTokenModifier = SemanticTokenModifier::new("documentation");
+    pub const DECLARATION: SemanticTokenModifier = SemanticTokenModifier::new("declaration");
+    pub const DEFINITION: SemanticTokenModifier = SemanticTokenModifier::new("definition");
+    pub const REFERENCE: SemanticTokenModifier = SemanticTokenModifier::new("reference");
+    pub const STATIC: SemanticTokenModifier = SemanticTokenModifier::new("static");
+    pub const ABSTRACT: SemanticTokenModifier = SemanticTokenModifier::new("abstract");
+    pub const DEPRECATED: SemanticTokenModifier = SemanticTokenModifier::new("deprecated");
+    pub const ASYNC: SemanticTokenModifier = SemanticTokenModifier::new("async");
+    pub const VOLATILE: SemanticTokenModifier = SemanticTokenModifier::new("volatile");
+    pub const READONLY: SemanticTokenModifier = SemanticTokenModifier::new("readonly");
+
+    pub const fn new(tag: &'static str) -> Self {
+        SemanticTokenModifier(Cow::Borrowed(tag))
+    }
+
+    pub fn as_str(&self) -> &str {
+        &self.0
+    }
+}
+
+#[cfg(feature = "proposed")]
+impl From<String> for SemanticTokenModifier {
+    fn from(from: String) -> Self {
+        SemanticTokenModifier(Cow::from(from))
+    }
+}
+#[cfg(feature = "proposed")]
+impl From<&'static str> for SemanticTokenModifier {
+    fn from(from: &'static str) -> Self {
+        SemanticTokenModifier::new(from)
+    }
+}
+
+/**
+ * @since 3.16.0 - Proposed state
+ */
+#[derive(Debug, Eq, PartialEq, Clone, Default, Deserialize, Serialize)]
+#[serde(rename_all = "camelCase")]
+#[cfg(feature = "proposed")]
+pub struct SemanticTokensLegend {
+    /**
+     * The token types a server uses.
+     */
+    pub token_types: Vec<SemanticTokenType>,
+
+    /**
+     * The token modifiers a server uses.
+     */
+    pub token_modifiers: Vec<SemanticTokenModifier>,
+}
+
+/**
+ * @since 3.16.0 - Proposed state
+ */
+#[derive(Debug, Eq, PartialEq, Clone, Default, Deserialize, Serialize)]
+#[serde(rename_all = "camelCase")]
+#[cfg(feature = "proposed")]
+pub struct SemanticTokens {
+    /**
+     * An optional result id. If provided and clients support delta updating
+     * the client will include the result id in the next semantic token request.
+     * A server can then instead of computing all sematic tokens again simply
+     * send a delta.
+     */
+    #[serde(skip_serializing_if = "Option::is_none")]
+    pub result_id: Option<String>,
+
+    /**
+     * The actual tokens. For a detailed description about how the data is
+     * structured pls see
+     * https://github.com/microsoft/vscode-extension-samples/blob/5ae1f7787122812dcc84e37427ca90af5ee09f14/semantic-tokens-sample/vscode.proposed.d.ts#L71
+     */
+    pub data: Vec<u32>,
```

Comment:
As a language with value types, we might go for something more structured here and elsewhere, like

```rust
struct SemanticToken {
  delta_line: u32,
  delta_start: u32,
  length: u32,
  token_type: u32,
  token_modifiers_bitset: u32
}

pub struct SemanticTokens { 
  #[serde(serialize/deserialize with = ...)]
  pub data: Vec<SemanticToken>,
}
```

No strong opinion about whether we should.

Reply:
Hm this seems reasonable. Can we pass context down to custom serializers? We would need to pass in the `SemanticTokensLegend` (or some other type) that is the intersection of supported server/client tokens and modifiers.

Reply:
>We would need to pass in the SemanticTokensLegend (or some other type) that is the intersection of supported server/client tokens and modifiers.

Why? I suggest keeping it simple and *just* use `u32` as in the above comment. We won't need any context then?

Reply:
Ah right. I was getting ahead of myself.

Reply:
Okay implemented serialization/deserialized. We will probably need to do something similar for edits.

I find it a little weird that tokens are defined by a tuple of 5 integers in the comments but as far as the LSP is concerned they only represented relative to each other. I have a feeling we'll have a local (ex. `rust-analyzer`) `SemanticToken` type that is something like `(line, char, length, type, properties)`.

Reply:
Though in thinking about it more that's okay because as an example in rust-analyzer it will be something like `(TextRange, Type, Modifer)` as the local type.